pkg/printers/template_test.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package printers
+
+import (
+	"bytes"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	api "k8s.io/kubernetes/pkg/apis/core"
+)
+
+func TestTemplate(t *testing.T) {
+	testCase := map[string]struct {
+		template  string
+		obj       runtime.Object
+		expectOut string
+		expectErr string
+	}{
+		"support base64 decoding of secret data": {
+			template: "{{ .Data.username | b64dec }}",
+			obj: &api.Secret{
+				Data: map[string][]byte{
+					"username": []byte("hunter"),
+				},
+			},
+			expectOut: "hunter",
+		},
+		"test error path for base64 decoding": {
+			template:  "{{ .Data.username | b64dec }}",
+			obj:       &badlyMarshaledSecret{},
+			expectErr: "error executing template \"{{ .Data.username | b64dec }}\": template: output:1:20: executing \"output\" at <b64dec>: error calling b64dec: illegal base64 data at input byte 0",
+		},
+	}
+	for name, test := range testCase {
+		buffer := &bytes.Buffer{}
+		p, err := NewTemplatePrinter([]byte(test.template))
+		if err != nil {
+			if err.Error() != test.expectErr {
+				t.Errorf("[%s]expect:\n %v\n but got:\n %v\n", name, test.expectErr, err)
+			}
+		} else {
+			err = p.PrintObj(test.obj, buffer)
+		}
+		if err != nil {
```

Comment:
Because of

```
err = p.PrintObj(test.obj, buffer)
```

`err` can never be `nil`

Also please avoid `else if`. For example

```
if err != nil && err.Error() != test.expectErr {
    t.Errorf("[%s]expect:\n %v\n but got:\n %v\n", name, test.expectErr, err)
    continue
}
if test.expectErr != "" {
    t.Errorf("[%s]expect:\n error %v\n but got:\n no error\n", name, test.expectErr)
}
```

Reply:
Thanks @ericchiang! I'm not sure about the first comment -- `p.PrintObj(test.obj, buffer)` will return `nil` if there is no error?

I'll try to apply the second comment tomorrow; I am having some difficulty unwinding the logic I've got into the structure you requested without breaking the tests.

Reply:
Just had to walk away for a couple of minutes, tests are passing now with simplified structure. Thanks for the comment, this looks a lot clearer now!